### PR TITLE
Set up a userspace stack before jumping to the payload.

### PR DESCRIPTION
**Pull request by andrisaar**
Another step closer to Ring 3: this moves the actual userspace stack into the lower half of the memory, away from the kernel space.

### CODE REVIEW

File: oak_restricted_kernel/src/payload.rs [hunk 1]
```diff
@@ -71,12 +71,25 @@ pub unsafe fn run_payload(payload: *const u8) -> ! {
             phdr.p_filesz as usize,
         );
     }
+    // Set up the userspace heap at the end of the lower half of the virtual address space.
```
Comment (conradgrobler) [inline oak_restricted_kernel/src/payload.rs:74]: "stack" rather than "heap"?

File: oak_restricted_kernel/src/payload.rs [hunk 2]
```diff
@@ -71,12 +71,25 @@ pub unsafe fn run_payload(payload: *const u8) -> ! {
...
+    // Well... almost. It's one page lower than the very end, as otherwise the initial stack pointer
+    // would need to be a noncanonical address, so let's avoid that whole mess by moving down a bit.
+    let rsp = VirtAddr::new(0x8000_0000_0000 - Size2MiB::SIZE);
+    mmap(
+        Some(rsp - Size2MiB::SIZE),
```
Comment (conradgrobler) [inline oak_restricted_kernel/src/payload.rs:79]: I assume 2MiB has been big enough for the stack for the stuff we are running so far. Is there a plan to be able to grow this in future up to some limit? (e.g. a page fault handler that will add more pages on demand)

**Review by conradgrobler — APPROVED** [review]
NIce, getting pretty close to a separate user-space.

File: oak_restricted_kernel/src/payload.rs (hunk 1 above)
Comment (andrisaar) [reply oak_restricted_kernel/src/payload.rs:74]: Yes, obviously.

File: oak_restricted_kernel/src/payload.rs (hunk 2 above)
Comment (andrisaar) [reply oak_restricted_kernel/src/payload.rs:79]: 2M should indeed be enough for now. There is no mechanism to extend the stack right now.

If we want to increase the stack size in the future, there are two options:
a) Just allocate more memory here. I think Linux by default allocates 8 MB for stacks. Obviously this comes at the cost of potentially wasting memory.
b) Do on-demand allocation (which we have discussed before): add entries to the page table, but don't allocate physical frames (yet). Once we get a page fault, only then allocate physical memory to back it. The benefit there is that even if we say that a stack is 8 MB, it won't really consume physical memory unless we need that much.

In any case we can punt that discussion down the line to the time when we actually need to figure it out.

**Review by ipetr0v — APPROVED** [review]

**Pull request merged**